### DISCUSSION TITLE
Set env variable value using task option

### DISCUSSION THREAD

**Discussion by pklejch** [nat-n/poethepoet#248] [Q&A] [answered]
Hello,

I have a 3rd party tool which only respects the `AWS_PROFILE` env variable. I need to set this env variable using poe before running this tool. Also I would like to support both `AWS_PROFILE` and `--profile <profile>` option for our poe wrapper.

tldr:
1. When `AWS_PROFILE` is set -> just pass it further -> done
2. If `AWS_PROFILE` is not set, use `--profile <profile>` and set the value of env variable `AWS_PROFILE` with value from the option.

Interestingly this works using this example
```
[tool.poe.tasks.my_task]
help = "Run my task"
control.expr = "'my_task'"

    [[tool.poe.tasks.my_task.switch]]
    # this is intentionally switch task with single case, because otherwise env AWS_PROFILE is not set
    case = "my_task"
    cmd = 'echo running task with ${profile}, env is $AWS_PROFILE'
    env = { AWS_PROFILE.default = "${profile}" }

    [[tool.poe.tasks.my_task.args]]
    name = "profile"
    positional = false
    help = "Which AWS profile to use. If no profile value is provided AWS_PROFILE env var will be used as default."
    default = "${AWS_PROFILE}"
```
notice the "single case switch".

And it works as I would like
```
$ poe my_task --profile a
Poe <= 'my_task'
Poe => echo running task with a, env is a
running task with a, env is a


$ AWS_PROFILE=b poe my_task
Poe <= 'my_task'
Poe => echo running task with b, env is b
running task with b, env is b
```
Both use cases are covered ✅ 

But when I try to simplify the poe configuration using this
```
[tool.poe.tasks.my_task]
help = "Run my task"
cmd = 'echo running task with ${profile}, env is ${AWS_PROFILE}'
env = { AWS_PROFILE.default = "${profile}" }

    [[tool.poe.tasks.my_task.args]]
    name = "profile"
    positional = false
    help = "Which AWS profile to use. If no profile value is provided AWS_PROFILE env var will be used as default."
    default = "${AWS_PROFILE}"
```
it does not work.
```
$ poe my_task --profile a
Poe => echo running task with a, env is
running task with a, env is

$ AWS_PROFILE=b poe my_task
Poe => echo running task with b, env is b
running task with b, env is b
```
Notice the empty `AWS_PROFILE` env variable when using the the `--profile` option.

Any tips how to use the simplified version ? Or it has something to do with subprocesses and passing env variables and it cannot be easily achieved otherwise ?

**Comment by nat-n** [comment, marked as answer]
Hi @pklejch, I just got around to looking at this properly and now I see that I misunderstood your problem the first time, so I deleted my original answer.

I understand you want your task to set AWS_PROFILE based on the `--profile` arg with a fallback to inheriting it from the environment. The approach is reasonable doesn't work because as it happens poe resolves the env for the task from other options before parsing the task arguments (so a CLI argument could have a default value based on the env task option, but not vice versa).

However, since arg values are actually set as environment variables, you can still get what you what simply by renaming the arg like so:

```toml
[tool.poe.tasks.my_task]
cmd  = 'echo "Running with profile: ${AWS_PROFILE}"'
args = [{ name = "AWS_PROFILE", options = ["--profile"], default = "${AWS_PROFILE}" }]
```

**Reply by pklejch** [reply]
Sorry, for the late response. 
I wanted to reply to you and reformulate the question, so it is more clear... But you answered me anyway !

Your solution works exactly as I wanted, thank you so much !
